### DISCUSSION TITLE
Ability to have SensorEntities or BinaryEntities as parameters to read from YAML for an App

### DISCUSSION THREAD

**Discussion by ngelotte** [net-daemon/netdaemon#307] [Feature requests]
It would be nice to not have to leave the GeneratedEntity world with all of it's code niceties in order to have the app be flexible and configurable from YAML.  So I am proposing being able to specify a parameter like
`public  SwitchEntity? EntityToControl{ get; set; }`

Netdaemon on load would bind the switch entity based on the string in the YAML file. Then in code you could use EntityToControl like any other generated Entity. It would be really nice to be able to have multiple entities in a list like:
`public  List<SwitchEntity>? EntityToControl{ get; set; }`

So that you could have multiple lights or sensors that you are referencing. I have not though this through as much so this might introduce some problems. I am not sure.

**Comment by suddenelfilio** [comment]
So the add to the discussion it would be nice to include the now generated specific entity types by default in netdaemon and extend the YamlAppconfig to allow binding the yaml value to the specific-entity typed property.

**Reply by suddenelfilio** [reply]
or more precise add support for RXEntity creation in the ToObject of the YamlExtensions class if I am correct

**Reply by suddenelfilio** [reply]
![image](https://user-images.githubusercontent.com/6460448/106664951-03f07980-65a6-11eb-8edc-caa0246b9350.png)

this is locally by some minor changes to the YamlExtensions and YamlConfig ;-) I'll have a look to poor this into a PR

**Reply by suddenelfilio** [reply]
Got a preliminary version already up-and-running. Still need to write some tests for it though, have a look: https://github.com/suddenelfilio/netdaemon/tree/features/307-Ability_to_have_SensorEntities_or_BinaryEntities_as_parameters_to_read_from_YAML_for_an_App

**Reply by ngelotte** [reply]
The problem to sort out is that because it is reading from HomeAssistant Entities will be different for different people. For example
in your ScriptEntity.cs you have a method for StopSonosSpeaker. My Instance won't have that because I don't have any SonosSpeakers. Mine has a method called RefillCurrentBucket. 

I think this would need to be just for the most basic entities like BinanySensor, Switch, etc that are universal to HomeAssistant.

**Reply by suddenelfilio** [reply]
Yes you are right. Might need to go through them all and just keep de default stuff and get rid of the custom methods specific to my install. This will mean the CodeGenerator needs to be adapted as well to exclude everything we keep and add the custom methods for a user's HASS install. That will be more involved though.

**Reply by helto4real** [reply]
If we do not need to generate them if we want to make them from yaml. We could create classes for each common type of entity that will be used by domain and use this in binding to the app property. This will be two different aproaches. I like this idea though.
